File: solution1.py
# https://leetcode.com/problems/delete-and-earn

# // Time Complexity : o(max_value)
# // Space Complexity : o(max_value)
# // Did this code successfully run on Leetcode :yes
# // Any problem you faced while coding this : None
#
#
# // Your code here along with comments explaining your approach

class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        if len(nums) == 0 or nums is None:
            return 0
        max_value = max(nums)
        list1 = [0 for i in range(max_value + 1)]
        for i in range(len(nums)):
            list1[nums[i]] += nums[i]

        dp = [[0 for i in range(2)] for j in range(max_value + 1)]

        for i in range(1, max_value + 1):
            for j in range(2):
                # include
                if j == 0:
                    dp[i][0] = list1[i] + dp[i - 1][1]
                # exclude:
                else:
                    dp[i][1] = max(dp[i - 1][0], dp[i - 1][1])
        return max(dp[max_value][0], dp[max_value][1])

File: Solution1.java
// Leetcode Problem : https://leetcode.com/problems/delete-and-earn/description/
// Time Complexity : O(n) where n is the maximum integer value present in given array
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
/*
 * I create an array of size maximum number in given array + 1 and store the frequency of all numbers appearing in the array.
 * I set previous as 0 at first and current as the first element of the frequency array. I traverse from 1 to length of 
 * frequency array and set count[i] as Math.max(prev+count[i],curr). I then update previous to current and current to count[i].
 * At the end, I return count[count.length-1].
 */

class Solution {
    public int deleteAndEarn(int[] nums) {
        if(nums.length==1)
        {
            return nums[0];
        }
        int max = Integer.MIN_VALUE;
        
        for(int i=0;i<nums.length;i++)
        {
            max=Math.max(nums[i],max);
        }
        int count[] = new int[max+1];
        
        for(int i=0;i<nums.length;i++)
        {
            count[nums[i]]+=nums[i];
        }
        int prev = 0, curr = count[0];
        for(int i=1;i<count.length;i++)
        {
            count[i]=Math.max(prev+count[i],curr);
            prev=curr;
            curr=count[i];
        }
        return count[count.length-1];
    }
}